### ISSUE TITLE
Latest Google Adapter not compatible with QnA Maker or Luis

### ISSUE THREAD

**Issue by twardpragmaticworks** [BotBuilderCommunity/botbuilder-community-dotnet#263]
Configuration:
    <PackageReference Include="Microsoft.Bot.Builder" Version="4.9.0" />
    <PackageReference Include="Microsoft.Bot.Builder.AI.Luis" Version="4.8.0" />
    <PackageReference Include="Microsoft.Bot.Builder.AI.QnA" Version="4.8.0" />
    <PackageReference Include="Bot.Builder.Community.Adapters.Google" Version="4.8.6/.7" />

Reverting to 4.6.4 resolves.  Was receiving null reference exception when analyzing turn context - via Luis or QnA Maker - was happening exclusively on the Google channel.

**Comment by garypretty** [comment]
Can you please provide a little more detail about the issue you are facing?  Specifically what do you mean that the adapter is not compatible with LUIS or QnA Maker? There have been several changes to the Google adapter recently, so we need to identify what exactly is happening.

**Comment by garypretty** [comment]
You can find the latest docs for the Google adapter at https://github.com/BotBuilderCommunity/botbuilder-community-dotnet/tree/develop/libraries/Bot.Builder.Community.Adapters.Google.  It would be great if you can compare your current configuration to this to identify where there might be a change.  Please let us know if you identify it so we can add a sign post to the docs or make a fix in the adapter.

**Comment by twardpragmaticworks** [comment]
@garypretty 
When passing the stepContext to the Luis Dispatcher or QnA Maker (The Microsoft Nuget packages for each respectively), a null object reference exception was returned.   My configuration is identical aside from the version. 

See code below: 

````
            var recognizerResult = await _botServices.Dispatch.RecognizeAsync(stepContext.Context, cancellationToken);

`````

**Comment by garypretty** [comment]
@twardpragmaticworks Can you confirm what the stepContext.Context.Activity.Text value is please before you pass in the context to the recognizer?

**Comment by twardpragmaticworks** [comment]
StepContext.Context.Activity.Text returns the utterance as expected, StepContext.Result however returned null.

**Comment by AndyfromPBG** [comment]
For what it's worth, I'm getting the same error. I followed the guide that Gary linked to. Thanks so much for your help, I'm looking into it myself and not just waiting on a handout. Here's my stack trace:
   at Microsoft.Bot.Builder.AI.Luis.LuisRecognizer.FillLuisEventPropertiesAsync(RecognizerResult recognizerResult, ITurnContext turnContext, Dictionary`2 telemetryProperties, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.AI.Luis.LuisRecognizer.OnRecognizerResultAsync(RecognizerResult recognizerResult, ITurnContext turnContext, Dictionary`2 telemetryProperties, Dictionary`2 telemetryMetrics, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.AI.Luis.LuisRecognizer.RecognizeInternalAsync(ITurnContext turnContext, LuisRecognizerOptions predictionOptions, Dictionary`2 telemetryProperties, Dictionary`2 telemetryMetrics, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.AI.Luis.LuisRecognizer.RecognizeAsync[T](ITurnContext turnContext, CancellationToken cancellationToken)
   at BackOfficeScheduler.SchedulerBot.AppointmentBookingRecognizer.RecognizeAsync[T](ITurnContext turnContext, CancellationToken cancellationToken)

**Comment by garypretty** [comment]
Thanks for the stack trace. I have an idea now - I suspect this might be due to a null user Id on the activity, which is probably throwing an exception when logging telemetry. I'll repro this morning and if it is I will get an update out today (or at the very least a workaround for you until one is available).

**Comment by twardpragmaticworks** [comment]
@AndyfromPBG 4.6.4 Adapter works in the interim.

**Comment by AndyfromPBG** [comment]
@twardpragmaticworks Thanks for the info, it would have taken me a while to get all the way down there, ha. In case anybody else needs to revert the setup to that version, it looks like this is the right readme: https://github.com/BotBuilderCommunity/botbuilder-community-dotnet/blob/3dd51332c9986e21412e55f399918cb22dadab73/libraries/Bot.Builder.Community.Adapters.Google/README.md

**Comment by AndyfromPBG** [comment]
@garypretty Just a heads up, I'm finally getting around to this again. I'll download the source and have a look myself- thanks for the pointer. I'll report back with what I find.

**Comment by AndyfromPBG** [comment]
The problem is somewhere in luis, so unfortunately debugging wasn't as quick as I'd hoped. Just so I don't lose track and in case anybody else picks this up, here's json of the activity coming from google:

{"Type":"message","Id":"e4f20e6f-2933-4947-8a86-fe23325a141a","Timestamp":"2020-07-06T19:06:56.4613299+00:00","LocalTimestamp":null,"LocalTimezone":null,"ServiceUrl":null,"ChannelId":"google","From":null,"Conversation":{"IsGroup":false,"ConversationType":null,"Id":"ABwppHHXImkeSpYe02mb2qT4mGV9H_Yyzk68yY6juSHqVCy93tnkX13wFq9Hy4yNU4sVsSiIY3vnYorL_IIlzz7jz_LrtRk","Name":null,"AadObjectId":null,"Role":null,"TenantId":null,"Properties":{}},"Recipient":{"Id":"","Name":"action","AadObjectId":null,"Role":null,"Properties":{}},"TextFormat":null,"AttachmentLayout":null,"MembersAdded":null,"MembersRemoved":null,"ReactionsAdded":null,"ReactionsRemoved":null,"TopicName":null,"HistoryDisclosed":null,"Locale":"en-US","Text":"i\u0027d like a massage","Speak":null,"InputHint":null,"Summary":null,"SuggestedActions":null,"Attachments":null,"Entities":null,"ChannelData":{"IsInSandbox":true,"Surface":{"Capabilities":[{"Name":"actions.capability.SCREEN_OUTPUT"},{"Name":"actions.capability.ACCOUNT_LINKING"},{"Name":"actions.capability.AUDIO_OUTPUT"},{"Name":"actions.capability.MEDIA_RESPONSE_AUDIO"},{"Name":"actions.capability.WEB_BROWSER"}]},"Inputs":[{"RawInputs":[{"Query":"I\u0027d like a massage at the lane Spa","InputType":"VOICE","Url":null}],"Arguments":[{"RawText":"I\u0027d like a massage at the lane Spa","TextValue":"I\u0027d like a massage","Name":"text","Extension":null,"IntValue":0,"FloatValue":0,"BoolValue":false,"DateTimeValue":"0001-01-01T00:00:00","PlaceValue":null,"StructuredValue":null}],"Intent":"actions.intent.TEXT"}],"User":{"IdToken":null,"Profile":null,"AccessToken":null,"Locale":"en-US","LastSeen":"2020-07-06T19:06:29Z","UserStorage":null,"PackageEntitlements":null,"UserVerificationStatus":"VERIFIED","UserId":null},"Device":null,"Conversation":{"ConversationId":"ABwppHHXImkeSpYe02mb2qT4mGV9H_Yyzk68yY6juSHqVCy93tnkX13wFq9Hy4yNU4sVsSiIY3vnYorL_IIlzz7jz_LrtRk","Type":"NEW","ConversationToken":null},"AvailableSurfaces":null},"Action":null,"ReplyToId":null,"Label":null,"ValueType":null,"Value":null,"Name":null,"RelatesTo":null,"Code":null,"Expiration":null,"Importance":null,"DeliveryMode":"expectReplies","ListenFor":null,"TextHighlights":null,"SemanticAction":null,"CallerId":null,"Properties":{}}

**Comment by garypretty** [comment]
I don't see anything wrong with the activity, other than some odd characters in the Text. Maybe you could try the same without the char?

Also, as an FYI, Google recently released a new version of the Actions SDK, which seemingly supersedes both the legacy Actions SDK and DialogFlow.  You can find the new adapter here -> https://github.com/BotBuilderCommunity/botbuilder-community-dotnet/tree/develop/libraries/Bot.Builder.Community.Adapters.ActionsSDK

**Comment by AndyfromPBG** [comment]
Odd, I got a null reference using the new adapter too. It must be something in how my code is configured, something in the bot initialization. I think I should probably start with the hello world and slowly introduce my bot code into it to find the problem.

   at Bot.Builder.Community.Adapters.ActionsSDK.Core.ActionsSdkRequestMapper.RequestToActivity(ActionsSdkRequest request)
   at Bot.Builder.Community.Adapters.ActionsSDK.ActionsSdkAdapter.ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken)
   at BackOfficeScheduler.SchedulerBot.Controllers.ActionsSdkController.PostAsync() in C:\BackOfficeScheduler\BackOfficeScheduler.SchedulerBot\Controllers\ActionsSdkController.cs:line 26

**Comment by garypretty** [comment]
@AndyfromPBG hey. This issue is referring to use of the legacy actions sdk / dialog flow adapter. Could you raise a separate issue for this using the new Actions SDK please?

I am keen to work with you to understand the issue here though as I want to ensure the new adapter is working.

Could you please share the json for the request from Google that is failing please? You should be able to get that from the simulator and it should tell me where it is failing.

Thanks

Gary

**Comment by AndyfromPBG** [comment]
@garypretty Sorry to get back to you on this so late- I updated all my references and don't get this problem anymore, so there's nothing to post to the Actions board- consider it closed. Thanks so much for your hard work and help.
